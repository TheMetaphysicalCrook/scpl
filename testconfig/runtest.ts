import * as fs from "fs";
import * as path from "path";
import * as prettyFormat from "pretty-format";

import { parse, inverse } from "..";

export function noUUID(
	obj: {},
	options: { noScPLData?: boolean; ignoreOutputName?: boolean } = {}
) {
	const uuids: string[] = [];
	return JSON.parse(
		JSON.stringify(obj, (key, value: unknown) => {
			if (options.noScPLData && key === "SCPLData") {
				return undefined;
			}
			if (typeof value === "string") {
				if (
					value.match(
						/[a-z0-9]{6}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}/
					)
				) {
					let index = uuids.indexOf(value);
					if (index === -1) {
						index = uuids.push(value) - 1; // push returns array length
					}
					return `<uuid${index + 1}>`;
				}
				if (
					options.ignoreOutputName &&
					(key === "CustomOutputName" || key === "OutputName")
				) {
					return undefined;
				}
				return value.split("\uFFFC").join("[attachment]");
			}
			return value;
		})
	);
}

export function runTest(dirname: string) {
	const infiles = fs.readdirSync(dirname);
	infiles.forEach(infile => {
		if (!infile.endsWith(".scpl")) {
			return;
		}
		test(infile, () => {
			const infilepath = path.join(dirname, infile);
			const contents = fs.readFileSync(infilepath, "utf-8");
			const [shortcut, expected] = contents.split(
				"\n\n@!ShouldEqual --------------------------------\n\n"
			);

			const got = [];

			let parseoutput: any;
			try {
				parseoutput = parse(shortcut, {
					make: ["shortcutjson"],
					useWarnings: true
				});
			} catch (e) {
				got.push(`Output Error:\n${prettyFormat(e)}`);
			}
			let pass: boolean = true;
			let _check1: any;
			let _equals1: any;
			if (parseoutput) {
				const output = parseoutput.shortcutjson;
				const warnings = parseoutput.warnings;

				got.push(`Output Warnings:\n${prettyFormat(warnings)}`);

				if (warnings.length === 0) {
					// invert
					const inverted = inverse(output);

					got.push(`Shortcut Full Inverted:\n${inverted}`);

					let reparsedoutput;
					let parsed;

					try {
						reparsedoutput = parse(inverted, {
							make: ["shortcutjson"],
							useWarnings: true
						});
						parsed = reparsedoutput.shortcutjson;
					} catch (e) {
						got.push(
							`!!! Error Invert Parsing ${
								(<Error>e).message
							} ${` ${(<Error>e).stack}`}`
						);
						pass = false;
					}

					got.push(
						`Shortcut Full JSON:\n${prettyFormat(noUUID(output))}`
					);

					_check1 = noUUID(parsed, {
						noScPLData: true,
						ignoreOutputName: true
					});
					_equals1 = noUUID(output, {
						noScPLData: true,
						ignoreOutputName: true
					});
				} else {
					got.push(
						"Because there were warnings, no inversion test will be run!!"
					);

					got.push(
						`Shortcut Full JSON:\n${prettyFormat(noUUID(output))}`
					);
				}
			}
			const gotString = got.join("\n\n");

			if (!expected || !expected.trim() || true) {
				const setValue = `${shortcut}\n\n@!ShouldEqual --------------------------------\n\n${gotString}`;
				if (contents === setValue) {
				} else {
					fs.writeFileSync(infilepath, setValue, "utf-8");
				}
			}

			if (_check1 || _equals1) {
				expect(_check1).toStrictEqual(_equals1);
			}
			if (!pass) {
				expect("Error invert parsing").toEqual("No errors");
			}
			expect(expected).toEqual(gotString);
		});
	});
}
