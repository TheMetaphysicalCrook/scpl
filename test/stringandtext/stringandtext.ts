import { runTest } from "../../testconfig/runtest";

runTest(__dirname);
