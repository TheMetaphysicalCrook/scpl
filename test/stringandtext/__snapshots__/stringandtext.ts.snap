// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stringandtext.scpl (test 0): inverted shortcut 1`] = `
"Text a
SetVariable a
GetVariable v:a
If input=Equals value=b -> mv:If
	Text b -> mv:Text
	GetVariable mv:If
	ShowResult \\"Error! This should never happen. \\\\(v:a) = \\\\(mv:Text).\\"
Otherwise
	Comment \\"Problem!\\"
End"
`;

exports[`stringandtext.scpl (test 0): shortcut data 1`] = `
Array [
  Object {
    "WFWorkflowActions": Array [
      Object {
        "SCPLData": Object {
          "Position": Object {
            "end": Array [
              1,
              16,
            ],
            "start": Array [
              1,
              1,
            ],
          },
        },
        "WFWorkflowActionIdentifier": "is.workflow.actions.gettext",
        "WFWorkflowActionParameters": Object {
          "WFTextActionText": "a",
        },
      },
      Object {
        "SCPLData": Object {
          "Position": Object {
            "end": Array [
              1,
              16,
            ],
            "start": Array [
              1,
              13,
            ],
          },
        },
        "WFWorkflowActionIdentifier": "is.workflow.actions.setvariable",
        "WFWorkflowActionParameters": Object {
          "WFVariableName": "a",
        },
      },
      Object {
        "SCPLData": Object {
          "Position": Object {
            "end": Array [
              2,
              8,
            ],
            "start": Array [
              2,
              5,
            ],
          },
        },
        "WFWorkflowActionIdentifier": "is.workflow.actions.getvariable",
        "WFWorkflowActionParameters": Object {
          "WFVariable": Object {
            "Value": Object {
              "Aggrandizements": Array [],
              "Type": "Variable",
              "VariableName": "a",
            },
            "WFSerializationType": "WFTextTokenAttachment",
          },
        },
      },
      Object {
        "SCPLData": Object {
          "Position": Object {
            "end": Array [
              2,
              19,
            ],
            "start": Array [
              2,
              1,
            ],
          },
        },
        "WFWorkflowActionIdentifier": "is.workflow.actions.conditional",
        "WFWorkflowActionParameters": Object {
          "GroupingIdentifier": "<uuid1>",
          "UUID": "<uuid2>",
          "WFCondition": "Equals",
          "WFConditionalActionString": "b",
          "WFControlFlowMode": 0,
        },
      },
      Object {
        "SCPLData": Object {
          "Position": Object {
            "end": Array [
              3,
              66,
            ],
            "start": Array [
              3,
              58,
            ],
          },
        },
        "WFWorkflowActionIdentifier": "is.workflow.actions.gettext",
        "WFWorkflowActionParameters": Object {
          "UUID": "<uuid3>",
          "WFTextActionText": "b",
        },
      },
      Object {
        "SCPLData": Object {
          "Position": Object {
            "end": Array [
              2,
              19,
            ],
            "start": Array [
              2,
              1,
            ],
          },
        },
        "WFWorkflowActionIdentifier": "is.workflow.actions.getvariable",
        "WFWorkflowActionParameters": Object {
          "WFVariable": Object {
            "Value": Object {
              "Aggrandizements": Array [],
              "OutputName": "If",
              "OutputUUID": "<uuid2>",
              "Type": "ActionOutput",
            },
            "WFSerializationType": "WFTextTokenAttachment",
          },
        },
      },
      Object {
        "SCPLData": Object {
          "Position": Object {
            "end": Array [
              3,
              69,
            ],
            "start": Array [
              3,
              2,
            ],
          },
        },
        "WFWorkflowActionIdentifier": "is.workflow.actions.showresult",
        "WFWorkflowActionParameters": Object {
          "Text": Object {
            "Value": Object {
              "attachmentsByRange": Object {
                "{33, 1}": Object {
                  "Aggrandizements": Array [],
                  "Type": "Variable",
                  "VariableName": "a",
                },
                "{37, 1}": Object {
                  "Aggrandizements": Array [],
                  "OutputName": "Text",
                  "OutputUUID": "<uuid3>",
                  "Type": "ActionOutput",
                },
              },
              "string": "Error! This should never happen. [attachment] = [attachment].",
            },
            "WFSerializationType": "WFTextTokenString",
          },
        },
      },
      Object {
        "SCPLData": Object {
          "Position": Object {
            "end": Array [
              4,
              10,
            ],
            "start": Array [
              4,
              1,
            ],
          },
        },
        "WFWorkflowActionIdentifier": "is.workflow.actions.conditional",
        "WFWorkflowActionParameters": Object {
          "GroupingIdentifier": "<uuid1>",
          "WFControlFlowMode": 1,
        },
      },
      Object {
        "SCPLData": Object {
          "Position": Object {
            "end": Array [
              5,
              20,
            ],
            "start": Array [
              5,
              2,
            ],
          },
        },
        "WFWorkflowActionIdentifier": "is.workflow.actions.comment",
        "WFWorkflowActionParameters": Object {
          "WFCommentActionText": "Problem!",
        },
      },
      Object {
        "SCPLData": Object {
          "Position": Object {
            "end": Array [
              6,
              4,
            ],
            "start": Array [
              6,
              1,
            ],
          },
        },
        "WFWorkflowActionIdentifier": "is.workflow.actions.conditional",
        "WFWorkflowActionParameters": Object {
          "GroupingIdentifier": "<uuid1>",
          "WFControlFlowMode": 2,
        },
      },
    ],
    "WFWorkflowClientRelese": "2.1.2",
    "WFWorkflowClientVersion": "754",
    "WFWorkflowIcon": Object {
      "WFWorkflowIconGlyphNumber": 59511,
      "WFWorkflowIconImageData": Object {
        "data": Array [],
        "type": "Buffer",
      },
      "WFWorkflowIconStartColor": 2071128575,
    },
    "WFWorkflowInputContentItemClasses": Array [
      "WFAppStoreAppContentItem",
      "WFArticleContentItem",
      "WFContactContentItem",
      "WFDateContentItem",
      "WFEmailAddressContentItem",
      "WFGenericFileContentItem",
      "WFImageContentItem",
      "WFiTunesProductContentItem",
      "WFLocationContentItem",
      "WFDCMapsLinkContentItem",
      "WFAVAssetContentItem",
      "WFPDFContentItem",
      "WFPhoneNumberContentItem",
      "WFRichTextContentItem",
      "WFSafariWebPageContentItem",
      "WFStringContentItem",
      "WFURLContentItem",
    ],
    "WFWorkflowMinimumClientVersion": 411,
    "WFWorkflowTypes": Array [
      "NCWidget",
      "WatchKit",
    ],
  },
]
`;

exports[`stringandtext.scpl (test 0): warnings 1`] = `Array []`;

exports[`stringandtext.scpl (test 1): error 1`] = `[Error: Error from 3,16 to 3,17: Did you mean \`\\\\\`? The character \`e\` is not a valid escape sequence. See the docs page on string escapes for more info.]`;

exports[`stringandtext.scpl (test 2): inverted shortcut 1`] = `"Comment \\"Interpolation where  isn't allowed\\""`;

exports[`stringandtext.scpl (test 2): shortcut data 1`] = `
Array [
  Object {
    "WFWorkflowActions": Array [
      Object {
        "SCPLData": Object {
          "Position": Object {
            "end": Array [
              3,
              58,
            ],
            "start": Array [
              3,
              1,
            ],
          },
        },
        "WFWorkflowActionIdentifier": "is.workflow.actions.comment",
        "WFWorkflowActionParameters": Object {
          "WFCommentActionText": "Interpolation where  isn't allowed",
        },
      },
    ],
    "WFWorkflowClientRelese": "2.1.2",
    "WFWorkflowClientVersion": "754",
    "WFWorkflowIcon": Object {
      "WFWorkflowIconGlyphNumber": 59511,
      "WFWorkflowIconImageData": Object {
        "data": Array [],
        "type": "Buffer",
      },
      "WFWorkflowIconStartColor": 2071128575,
    },
    "WFWorkflowInputContentItemClasses": Array [
      "WFAppStoreAppContentItem",
      "WFArticleContentItem",
      "WFContactContentItem",
      "WFDateContentItem",
      "WFEmailAddressContentItem",
      "WFGenericFileContentItem",
      "WFImageContentItem",
      "WFiTunesProductContentItem",
      "WFLocationContentItem",
      "WFDCMapsLinkContentItem",
      "WFAVAssetContentItem",
      "WFPDFContentItem",
      "WFPhoneNumberContentItem",
      "WFRichTextContentItem",
      "WFSafariWebPageContentItem",
      "WFStringContentItem",
      "WFURLContentItem",
    ],
    "WFWorkflowMinimumClientVersion": 411,
    "WFWorkflowTypes": Array [
      "NCWidget",
      "WatchKit",
    ],
  },
]
`;

exports[`stringandtext.scpl (test 2): warnings 1`] = `
Array [
  [Error: Error from 3,32 to 3,42: This string is not allowed to have variables.],
]
`;
